src/resolver.js

```diff
@@ -1,16 +1,152 @@
 'use strict'
 
 const util = require('./util')
-const setImmediate = require('async/setImmediate')
+const traverse = require('traverse')
 
 exports = module.exports
 
-exports.multicodec = 'git'
+exports.multicodec = 'git-raw'
+
+const personInfoPaths = [
+  'original',
+  'name',
+  'email',
+  'date'
+]
 
 exports.resolve = (block, path, callback) => {
-    setImmediate(() => callback(new Error("not implemented"), null))
+  if (typeof path === 'function') {
+    callback = path
+    path = undefined
+  }
+
+  util.deserialize(block.data, (err, node) => {
+    if (err) {
+      return callback(err)
+    }
+
+    if (!path || path === '/') {
+      return callback(null, {
+        value: node,
+        remainderPath: ''
+      })
+    }
+
+    if (Buffer.isBuffer(node)) { // git blob
+      return callback(null, {
+        value: node,
+        remainderPath: path
+      })
+    }
+
+    const parts = path.split('/')
+    const val = traverse(node).get(parts)
+
+    if (val) {
+      return callback(null, {
+        value: val,
+        remainderPath: ''
+      })
+    }
+
+    let value
+    let len = parts.length
+
+    for (let i = 0; i < len; i++) {
+      const partialPath = parts.shift()
+
+      if (Array.isArray(node)) {
+        value = node[Number(partialPath)]
+      } if (node[partialPath]) {
+        value = node[partialPath]
+      } else {
+        // can't traverse more
+        if (!value) {
+          return callback(new Error('path not available at root'))
+        } else {
+          parts.unshift(partialPath)
+          return callback(null, {
+            value: value,
+            remainderPath: parts.join('/')
+          })
+        }
+      }
+      node = value
+    }
+  })
 }
 
 exports.tree = (block, options, callback) => {
-    setImmediate(() => callback(new Error("not implemented"), null))
+  if (typeof options === 'function') {
+    callback = options
+    options = undefined
+  }
+
+  options = options || {}
+
+  util.deserialize(block.data, (err, node) => {
+    if (err) {
+      return callback(err)
+    }
+
+    if (Buffer.isBuffer(node)) { // git blob
+      return callback(null, [])
+    }
+
+    let paths = []
+    switch (node.gitType) {
+      case 'commit':
+        paths = [
+          'message',
+          'tree'
+        ]
+
+        paths = paths.concat(personInfoPaths.map(e => 'author/' + e))
+        paths = paths.concat(personInfoPaths.map(e => 'committer/' + e))
+        paths = paths.concat([...node.parents.keys()].map(e => 'parents/' + e))
+
+        if (node.encoding) {
+          paths.push('encoding')
+        }
+        break
+      case 'tag':
+        paths = [
+          'object',
+          'type',
+          'tag',
+          'message'
+        ]
```

Comment:
Are these constants? Move to a Constants file

Reply:
These few paths are there for every tag, I'm not really sure if it'd really be worth having a separate file for that.